### PR TITLE
mgr/cephadm: allow draining host without removing conf/keyring files

### PR DESCRIPTION
**Pull request by adk3798**
Fixes: https://tracker.ceph.com/issues/58820

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/inventory.py [hunk 1]
```diff
@@ -925,6 +925,24 @@ def get_schedulable_hosts(self) -> List[HostSpec]:
...
+        Any host without that label is considered fair game for
+        a client keyring spec to match. We want to still wait
+        for refresh here so that we know what keyrings we've
+        already deployed here
+        """
+        return [
+            h for h in self.mgr.inventory.all_specs()
+            if (
+                self.host_had_daemon_refresh(h.hostname)
+                and '_no_conf_keyring' not in h.labels
```
Comment (rkachach) [inline src/pybind/mgr/cephadm/inventory.py:942]: probably we need to declare a constant for `_no_conf_keyring` as it's used in several places

File: doc/cephadm/host-management.rst [hunk 2]
```diff
@@ -101,8 +101,19 @@ To drain all daemons from a host, run a command of the following form:
 
    ceph orch host drain *<host>*
 
-The ``_no_schedule`` label will be applied to the host. See
-:ref:`cephadm-special-host-labels`.
+The ``_no_schedule`` and ``_no_conf_keyring`` labels will be applied to the
+host. See :ref:`cephadm-special-host-labels`.
+
+If you only want to drain daemons but leave managed ceph conf and keyring
+files on the host, you may pass the ``--leave-conf-keyring`` flag to the
```
Comment (rkachach) [inline doc/cephadm/host-management.rst:108]: nit: what about using `--keep-conf-keyring` instead of `--leave-conf-keyring`?

File: doc/cephadm/host-management.rst [hunk 3]
```diff
@@ -183,6 +194,12 @@ The following host labels have a special meaning to cephadm.  All start with ``_
   an existing host that already contains Ceph daemons, it will cause cephadm to move
   those daemons elsewhere (except OSDs, which are not removed automatically).
 
+* ``_no_conf_keyring``: *Do not deploy config files or keyrings on this host*.
+
+  This label is effectively the same as ``_no_schedule`` but instead of working for
+  daemons it works for client keyrings and ceph conf files that are being managed
+  by cephadm
+
```
Comment (rkachach) [inline doc/cephadm/host-management.rst:202]: here I'm a little bit lost/confused: the `_no_conf_keyring` label is supposed to keep the configuration but as of this statment it seems to prevent deploying config on the host! .. I think we have to reword this to better reflect what's the purpose of this label.

File: src/pybind/mgr/cephadm/serve.py [hunk 4]
```diff
@@ -1087,9 +1087,9 @@ def _calc_client_files(self) -> Dict[str, Dict[str, Tuple[int, int, int, bytes,
                 pspec = PlacementSpec.from_string(self.mgr.manage_etc_ceph_ceph_conf_hosts)
                 ha = HostAssignment(
                     spec=ServiceSpec('mon', placement=pspec),
-                    hosts=self.mgr.cache.get_schedulable_hosts(),
+                    hosts=self.mgr.cache.get_conf_keyring_available_hosts(),
                     unreachable_hosts=self.mgr.cache.get_unreachable_hosts(),
-                    draining_hosts=self.mgr.cache.get_draining_hosts(),
+                    draining_hosts=self.mgr.cache.get_conf_keyring_draining_hosts(),
```
Comment (rkachach) [inline src/pybind/mgr/cephadm/serve.py:1092]: In the code there're more calls to `mgr.cache.get_schedulable_hosts()` but we only replaced these calls. Please, can you explain why we didn't do the same in all the calls (just for me to understand better the logic behind it).

Comment (adk3798) [reply src/pybind/mgr/cephadm/serve.py:1092]: most of those calls are for scheduling the placement of actual daemons. The reason I adjusted these two specific calls is because they are part of the `_calc_client_files` function that specifically is for generating placements for ceph conf and client keyrings. These placement calls, even though it says it's for a mon service, are actually for those client files.

File: doc/cephadm/host-management.rst (hunk 3 above)
Comment (adk3798) [reply doc/cephadm/host-management.rst:202]: `_no_conf_keyring` does in fact prevent cephadm from deploying the conf and keyring files on the host. I think the confusion is with the fact that the `--leave-conf-keyring` flag (or `--keep-conf-keyring` if I rename it as you suggested in another comment) tells the drain host function NOT to put this label on the host. To summarize:

- drain with NO `--leave-conf-keyring` flag -> Add `_no_conf_keyring` label and conf and keyring files get removed
- drain WITH `--leave-conf-keyring` flag -> do NOT add `_no_conf_keyring` label and the conf and keyring files are left as is.
- In both cases, add `_no_schedule` label that will cause daemons on the host to be removed

So in that sense the label does work exactly like `_no_schedule` but for conf and keyrings instead of daemons. It's just that we have an optional flag for the drain host function that tells us not to add the label as part of the drain process.

File: src/pybind/mgr/cephadm/inventory.py (hunk 1 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py]: didn't want to handle this one differently than the other special host labels, so I added a class in utils that defines them in a separate commit

Comment (rkachach) [reply src/pybind/mgr/cephadm/inventory.py]: sounds reasonable :+1:

File: doc/cephadm/host-management.rst (hunk 3 above)
Comment (rkachach) [reply doc/cephadm/host-management.rst:202]: Ok, I see thanks for the detailed explanation. Maybe the confusion in fact comes from the double negation usage but if you think about it in terms of how _no_schedule works then it's becomes easiser to understand.

File: src/pybind/mgr/cephadm/serve.py (hunk 4 above)
Comment (rkachach) [reply src/pybind/mgr/cephadm/serve.py:1098]: I see, that's right. I didn't see at first glance that it was called from `_calc_client_files` function.

File: src/pybind/mgr/cephadm/inventory.py [hunk 5]
```diff
@@ -946,6 +964,15 @@ def get_draining_hosts(self) -> List[HostSpec]:
...
 
+    def get_conf_keyring_draining_hosts(self) -> List[HostSpec]:
+        """
+        Returns all hosts that have _no_conf_keyring label and therefore should have
+        no config files or client keyring placed on them, but are potentially still reachable
+        """
+        return [
+            h for h in self.mgr.inventory.all_specs() if '_no_conf_keyring' in h.labels
+        ]
+
```
Comment (rkachach) [inline src/pybind/mgr/cephadm/inventory.py:975]: shouldn't this function check also for the `_no_schedule` label? the other `get_draining_hosts` function does check for this label. If it shouldn't then maybe we should remove the `draining` from the function name? WDYT

Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py:975]: no, the whole idea here is to separate out draining the daemons from a host with draining the conf and keyring to allow doing one but not the other. If we check `_no_schedule` here then the label that causes the daemons to be drained will also cause the conf keyrings to be drained which is how it worked before and what this PR is trying to change.

Comment (rkachach) [reply src/pybind/mgr/cephadm/inventory.py:975]: I see, I thought that `draining` in this context was refering to daemons as well. In this case we are just draining the conf.

**Comment by adk3798** [comment]
jenkins test make check

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-03-06_16:09:09-orch:cephadm-wip-adk-testing-2023-03-04-1231-distro-default-smithi/

Failed/dead jobs reruns: https://pulpito.ceph.com/adking-2023-03-06_21:04:02-orch:cephadm-wip-adk-testing-2023-03-04-1231-distro-default-smithi/

After reruns, 6 failed, 2 dead jobs:

- 2 dead jobs were teuthology/infra issues (`Could not get lock /var/lib/dpkg/lock-frontend`, `SSH connection to smithi038 was lost`)
- 1 job failed installing a kernel `[Errno 2] No such file or directory: '/var/cache/dnf/CentOS-BaseOS-7c3d8b2bb640713c/packages/kernel-4.18.0-448.el8.x86_64.rpm'` unrelated to our tests
- 2 instances of https://tracker.ceph.com/issues/58535
- 3 jobs failed due to changes made to `ceph nfs cluster create` command in another PR in the run

Overall, should only block merging for PRs related to nfs

**Comment by bluikko** [comment]
Excuse me, I'm not too familiar with the codebase. Would it be possible to have this same feature in the `orch host maintenance enter` command?

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-03-08_02:46:19-orch:cephadm-wip-adk-testing-2023-03-07-1710-distro-default-smithi/

Reruns of failed/dead jobs: https://pulpito.ceph.com/adking-2023-03-08_13:40:40-orch:cephadm-wip-adk-testing-2023-03-07-1710-distro-default-smithi/

After reruns, 1 failed and 2 dead jobs:

- both dead jobs were due to `Failed to download packages: qemu-img-15:6.2.0-28.module_el8.8.0+1257+0c3374ae.x86_64: Cannot download, all mirrors were already tried without success` which is not relevant to us (didn't want to have to do another set of reruns)
- 1 failed job was https://tracker.ceph.com/issues/58535

Overall, nothing in that run that would block merging

**Comment by adk3798** [comment]
> Excuse me, I'm not too familiar with the codebase. Would it be possible to have this same feature in the `orch host maintenance enter` command?

afaik putting a host in maintenance mode shouldn't cause us to remove client/config files to begin with. Have you had that happen on your cluster?

**Comment by bluikko** [comment]
> afaik putting a host in maintenance mode shouldn't cause us to remove client/config files to begin with.

I see. I expected that is happening since I've had multiple (4?) cases where those files have gone missing, a few times around the time when maintenance mode was entered.  
Nevermind then and thanks for your reply.

**Comment by adk3798** [comment]
> > afaik putting a host in maintenance mode shouldn't cause us to remove client/config files to begin with.
> 
> I see. I expected that is happening since I've had multiple (4?) cases where those files have gone missing, a few times around the time when maintenance mode was entered. Nevermind then and thanks for your reply.

hmm, perhaps you were hitting https://tracker.ceph.com/issues/57462 (the bug has been fixed but the fix hasn't landed in any actual releases yet) and then because the host was in maintenance cephadm never put them back? If you keep seeing it you could raise another tracker. It's definitely not supposed to remove the conf/keyring from hosts being put in maintenance mode.

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-03-15_01:34:51-orch:cephadm-wip-adk-testing-2023-03-14-1905-distro-default-smithi/

reruns of failed/dead jobs: https://pulpito.ceph.com/adking-2023-03-30_12:19:06-orch:cephadm-wip-adk-testing-2023-03-28-1716-distro-default-smithi/

After reruns, 6 failures:

- 2 failed staggered upgrade test, combination of https://tracker.ceph.com/issues/58535 plus the attempted fix not working
- 2 failed nfs-keepalive-only tests, newly added by a PR in the run, and not an issue in terms of merging any other PR
- 1 instance of https://tracker.ceph.com/issues/49287 (known issue)
- 1 failure in test_nfs task. The failure only happens in the test variation with the agent on (https://pulpito.ceph.com/adking-2023-03-29_11:57:55-orch:cephadm-wip-adk-testing-2023-03-28-1716-distro-default-smithi/7225304 passed) so I believe this is a timing thing related to the agent. Nothing to block merging on anything not specifically related to nfs.

File: src/pybind/mgr/cephadm/utils.py [hunk 6]
```diff
@@ -43,6 +43,13 @@ class ContainerInspectInfo(NamedTuple):
     repo_digests: Optional[List[str]]
 
 
+class SpecialHostLabels():
```
Comment (phlogistonjohn) [inline src/pybind/mgr/cephadm/utils.py:46]: I'm curious why this is not an enum.  Otherwise, +1 to using constants.

File: src/pybind/mgr/cephadm/inventory.py [hunk 7]
```diff
@@ -958,7 +958,25 @@ def get_schedulable_hosts(self) -> List[HostSpec]:
...
                 self.host_had_daemon_refresh(h.hostname)
-                and '_no_schedule' not in h.labels
+                and SpecialHostLabels.drain_daemons_label not in h.labels
+            )
+        ]
+
+    def get_conf_keyring_available_hosts(self) -> List[HostSpec]:
+        """
+        Returns all hosts without _no_conf_keyring label that
+        have had a refresh
```
Comment (rkachach) [inline src/pybind/mgr/cephadm/inventory.py:968]: I think the comment should be explicit about what hosts is getting. Something like:

`Returns all hosts eligible to deploy conf and keyring files (without _no_conf_keyring label)`

File: src/pybind/mgr/cephadm/inventory.py [hunk 8]
```diff
@@ -971,7 +989,7 @@ def get_non_draining_hosts(self) -> List[HostSpec]:
         no daemon refresh
         """
         return [
-            h for h in self.mgr.inventory.all_specs() if '_no_schedule' not in h.labels
```
Comment (rkachach) [inline src/pybind/mgr/cephadm/inventory.py:974]: nit: consider upadating the comment (we shouldn't make reference to the `_no_schedule` instead we should use the `SpecialHostLabels.drain_daemons_label`

File: src/pybind/mgr/cephadm/inventory.py [hunk 9]
```diff
@@ -980,7 +998,16 @@ def get_draining_hosts(self) -> List[HostSpec]:
         no daemons placed on them, but are potentially still reachable
         """
         return [
-            h for h in self.mgr.inventory.all_specs() if '_no_schedule' in h.labels
+            h for h in self.mgr.inventory.all_specs() if SpecialHostLabels.drain_daemons_label in h.labels
+        ]
+
+    def get_conf_keyring_draining_hosts(self) -> List[HostSpec]:
+        """
+        Returns all hosts that have _no_conf_keyring label and therefore should have
```
Comment (rkachach) [inline src/pybind/mgr/cephadm/inventory.py:1006]: nit: idem to the previous comment

**Review by rkachach — APPROVED** [review]
LGTM in general, just some additional minor comments.

File: src/pybind/mgr/cephadm/utils.py [hunk 10]
```diff
@@ -43,6 +43,13 @@ class ContainerInspectInfo(NamedTuple):
...
+    admin_label: str = '_admin'
+    no_autotune_memory_label: str = '_no_autotune_memory'
+    drain_daemons_label: str = '_no_schedule'
+    drain_conf_keyring_label: str = '_no_conf_keyring'
```
Comment (rkachach) [inline src/pybind/mgr/cephadm/utils.py:50]: nit: it sounds like a good place to document what every label used-for/does :)

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-04-04_02:06:27-orch:cephadm-wip-adk-testing-2023-04-03-1541-distro-default-smithi/

Reruns of failed/dead jobs:  https://pulpito.ceph.com/adking-2023-04-04_05:19:12-orch:cephadm-wip-adk-testing-2023-04-03-1541-distro-default-smithi/

After reruns, 6 failures. All 6 failures were on tests being added/adjusted by PRs in the run (2 keepalive-only nfs tests, 2 mon set crush location tests, 2 staggered upgrade tests) and those failures don't seem like they should block merging for any PRs in the run outside of those adjusting the tests.

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-04-20_12:02:37-orch:cephadm-wip-adk-testing-2023-04-20-0105-distro-default-smithi/

Reruns of all but upgrade-with-workload jobs (which are known to timeout after multiple hours currently, so reruns would waste a lot of resources): https://pulpito.ceph.com/adking-2023-04-22_17:05:01-orch:cephadm-wip-adk-testing-2023-04-20-0105-distro-default-smithi/

After reruns, 2 dead and 5 failed jobs:

- 2 dead jobs were infra failures: `"Failed to download metadata for repo 'CentOS-PowerTools'`
- 2 failed jobs on staggered upgrade tests, https://tracker.ceph.com/issues/58535 + fix for it not working yet
- 2 failed jobs on new keepalive-only test. It's a new test that doesn't work yet, shouldn't affect being able to merge other PRs
- 1 failed test deploying jeagar-tracing. Needs investigation, but doesn't seem related to any PRs in the run.

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-05-03_12:22:42-orch:cephadm-wip-adk-testing-2023-05-02-1906-distro-default-smithi/

5 failures:

- 2 failures on staggered upgrade test. Combination of https://tracker.ceph.com/issues/58535 plus the attempted fix not quite working yet
- 1 failure in the test_cephadm task. Related to a PR in the run changing the `cephadm version` command
- 1 failure deploying jaegar-tracing. Still need to investigate this, but it's not new to this run, so shouldn't block merging.
- 1 failure making an OSD in the mds upgrade sequence `ERROR: error creating empty object store in /var/lib/ceph/osd/ceph-5/: (5) Input/output error`. These pop up occasionally in hte lab, most likely as a result of us using split nvme devices for OSDs. Nothing that would block merging. 

Overall, failures should only block merging for the 2 PRs related to the staggered upgrade and cephadm version changes. Everything else should be fine.

File: src/pybind/mgr/cephadm/utils.py (hunk 10 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/utils.py]: with this PR adding docs for the `_no_conf_keyring` label, I think they should all be covered in https://docs.ceph.com/en/latest/cephadm/host-management/#special-host-labels. It's technically docs for users rather than devs, but the behavior outlined there is basically what I would say in any code comment here anyway.

File: src/pybind/mgr/cephadm/utils.py (hunk 6 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/utils.py]: made this class an enum/str so I didn't have to worry about type signatures for things like the `has_label` func that we've always passed strings before

File: src/pybind/mgr/cephadm/inventory.py (hunk 7 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py:1013]: added a sentence like that to the comment

**Review by phlogistonjohn — APPROVED** [review]
Seems OK to me

**Comment by adk3798** [comment]
jenkins test make check

**Comment by adk3798** [comment]
jenkins test dashboard cephadm
